src/main/java/com/midaswebserver/midasweb/services/UserServiceImp.java

```diff
@@ -0,0 +1,264 @@
+            log.error("update: passed username was null");
+            return false;
+        }
+        if (this.getUserById(user.getId()) == null) {
+            log.warn("update: D of user '{}' didn't match anything in database", user.getUsername());
+            return false;
+        }
+        log.debug("update: User '{}' has been updated", user.getUsername());
+        userRepo.save(user);
+        return true;
+    }
+
+    /**
+     * Adds a symbol to a user. This will create a symbol object linked to the user object
+     * which can be used to retrieve the saved ticker
+     * WIll return false if user or ticker is null
+     * if ticker is invalid > 4, return false
+     * if User doesn't exist, return false
+     * if ticker is already saved, return true
+     * if ticker is added, return true
+     * all else fails, return false
+     *
+     * @param user
+     * @param ticker like 'UEC'
+     * @return boolean based on the success of the transaction
+     */
+    @Transactional
+    @Override
+    public boolean addSymbolToUser(User user, String ticker) {
+        //validation
+        if (user == null || ticker == null) {
+            log.error("addSymbolToUser: user, '{}', or symbol, '{}', was found to be null", user, ticker);
+            return false;
+        }
+        if (ticker.getBytes().length > 4) {
+            log.error("addSymbolToUser: symbol, '{}', was longer than expected, 4 characters", ticker);
+            return false;
+        }
+        if (!userRepo.existsById(user.getId())) {
+            log.error("addSymbolToUser: user, '{}', with id, '{}', was not found in database", user.getUsername(), user.getId());
+            return false;
+        }
+        //logic
+        for (Symbol symbolSaved : user.getSymbol()) {
+            if (symbolSaved.getTicker().equalsIgnoreCase(ticker)) {
+                log.debug("addSymbolToUser: user, '{}', attempted to add ticker, '{}', that was added previously", user.getUsername(), ticker);
+                return true;
+            }
+        }
+        user.addTicker(ticker, user);
+        update(user);
+        for (Symbol symbolSaved : user.getSymbol()) {
+            if (symbolSaved.getTicker().equalsIgnoreCase(ticker)) {
+                log.debug("addSymbolToUser: user, '{}', added ticker , '{}', to the database", user.getUsername(), ticker);
+                return true;
+            }
+        }
+        log.error("addSymbolToUser: user, '{}', failed to add ticker , '{}', should have found ticker in database", user.getUsername(), ticker);
+        return false;
+    }
+
+    /**
+     * takes request and finds the user's Ip. Used in place of user id when logging
+     *
+     * @param request {@link HttpServletRequest} takes the servlet request received from a post method
+     * @return remote IP address
+     */
+    @Override
+    public String getClientIp(HttpServletRequest request) {
+        String remoteAddr = "";
+
+        if (request != null) {
```

Comment:
Great job in figuring out the correct business logic for determining the IP address from a web request!